{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: face_recognition in c:\\anaconda\\lib\\site-packages (1.3.0)\n",
      "Requirement already satisfied: opencv-python in c:\\anaconda\\lib\\site-packages (4.8.1.78)\n",
      "Requirement already satisfied: dlib in c:\\anaconda\\lib\\site-packages (19.24.2)\n",
      "Requirement already satisfied: numpy in c:\\anaconda\\lib\\site-packages (1.26.2)\n",
      "Requirement already satisfied: pillow in c:\\anaconda\\lib\\site-packages (10.0.1)\n",
      "Requirement already satisfied: matplotlib in c:\\anaconda\\lib\\site-packages (3.8.4)\n",
      "Requirement already satisfied: jupyter in c:\\anaconda\\lib\\site-packages (1.1.1)\n",
      "Requirement already satisfied: face-recognition-models>=0.3.0 in c:\\anaconda\\lib\\site-packages (from face_recognition) (0.3.0)\n",
      "Requirement already satisfied: Click>=6.0 in c:\\anaconda\\lib\\site-packages (from face_recognition) (8.1.7)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\anaconda\\lib\\site-packages (from matplotlib) (1.2.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\anaconda\\lib\\site-packages (from matplotlib) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\anaconda\\lib\\site-packages (from matplotlib) (4.51.0)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\anaconda\\lib\\site-packages (from matplotlib) (1.4.5)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\anaconda\\lib\\site-packages (from matplotlib) (23.1)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in c:\\anaconda\\lib\\site-packages (from matplotlib) (3.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in c:\\anaconda\\lib\\site-packages (from matplotlib) (2.8.2)\n",
      "Requirement already satisfied: notebook in c:\\anaconda\\lib\\site-packages (from jupyter) (6.5.4)\n",
      "Requirement already satisfied: jupyter-console in c:\\anaconda\\lib\\site-packages (from jupyter) (6.6.3)\n",
      "Requirement already satisfied: nbconvert in c:\\anaconda\\lib\\site-packages (from jupyter) (7.10.0)\n",
      "Requirement already satisfied: ipykernel in c:\\anaconda\\lib\\site-packages (from jupyter) (6.25.0)\n",
      "Requirement already satisfied: ipywidgets in c:\\anaconda\\lib\\site-packages (from jupyter) (8.1.5)\n",
      "Requirement already satisfied: jupyterlab in c:\\anaconda\\lib\\site-packages (from jupyter) (3.6.3)\n",
      "Requirement already satisfied: colorama in c:\\anaconda\\lib\\site-packages (from Click>=6.0->face_recognition) (0.4.6)\n",
      "Requirement already satisfied: six>=1.5 in c:\\anaconda\\lib\\site-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
      "Requirement already satisfied: comm>=0.1.1 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (0.2.2)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (1.6.7)\n",
      "Requirement already satisfied: ipython>=7.23.1 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (8.15.0)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (7.4.9)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (5.5.0)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (0.1.6)\n",
      "Requirement already satisfied: nest-asyncio in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (1.5.6)\n",
      "Requirement already satisfied: psutil in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (5.9.0)\n",
      "Requirement already satisfied: pyzmq>=20 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (23.2.0)\n",
      "Requirement already satisfied: tornado>=6.1 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (6.3.3)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in c:\\anaconda\\lib\\site-packages (from ipykernel->jupyter) (5.7.1)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.12 in c:\\anaconda\\lib\\site-packages (from ipywidgets->jupyter) (4.0.13)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0.12 in c:\\anaconda\\lib\\site-packages (from ipywidgets->jupyter) (3.0.13)\n",
      "Requirement already satisfied: prompt-toolkit>=3.0.30 in c:\\anaconda\\lib\\site-packages (from jupyter-console->jupyter) (3.0.36)\n",
      "Requirement already satisfied: pygments in c:\\anaconda\\lib\\site-packages (from jupyter-console->jupyter) (2.15.1)\n",
      "Requirement already satisfied: jupyterlab-server~=2.19 in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (2.25.1)\n",
      "Requirement already satisfied: jupyter-server<3,>=1.16.0 in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (1.23.4)\n",
      "Requirement already satisfied: jupyter-ydoc~=0.2.3 in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (0.2.4)\n",
      "Requirement already satisfied: jupyter-server-ydoc~=0.8.0 in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (0.8.0)\n",
      "Requirement already satisfied: nbclassic in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (1.0.0)\n",
      "Requirement already satisfied: jinja2>=2.1 in c:\\anaconda\\lib\\site-packages (from jupyterlab->jupyter) (3.1.2)\n",
      "Requirement already satisfied: argon2-cffi in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (21.3.0)\n",
      "Requirement already satisfied: ipython-genutils in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (0.2.0)\n",
      "Requirement already satisfied: nbformat in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (5.9.2)\n",
      "Requirement already satisfied: Send2Trash>=1.8.0 in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (1.8.2)\n",
      "Requirement already satisfied: terminado>=0.8.3 in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (0.17.1)\n",
      "Requirement already satisfied: prometheus-client in c:\\anaconda\\lib\\site-packages (from notebook->jupyter) (0.14.1)\n",
      "Requirement already satisfied: beautifulsoup4 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (4.12.2)\n",
      "Requirement already satisfied: bleach!=5.0.0 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (4.1.0)\n",
      "Requirement already satisfied: defusedxml in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (0.7.1)\n",
      "Requirement already satisfied: jupyterlab-pygments in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (0.1.2)\n",
      "Requirement already satisfied: markupsafe>=2.0 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (2.1.1)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (2.0.4)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (0.8.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in c:\\anaconda\\lib\\site-packages (from nbconvert->jupyter) (1.2.1)\n",
      "Requirement already satisfied: webencodings in c:\\anaconda\\lib\\site-packages (from bleach!=5.0.0->nbconvert->jupyter) (0.5.1)\n",
      "Requirement already satisfied: backcall in c:\\anaconda\\lib\\site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.2.0)\n",
      "Requirement already satisfied: decorator in c:\\anaconda\\lib\\site-packages (from ipython>=7.23.1->ipykernel->jupyter) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in c:\\anaconda\\lib\\site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.18.1)\n",
      "Requirement already satisfied: pickleshare in c:\\anaconda\\lib\\site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.7.5)\n",
      "Requirement already satisfied: stack-data in c:\\anaconda\\lib\\site-packages (from ipython>=7.23.1->ipykernel->jupyter) (0.2.0)\n",
      "Requirement already satisfied: entrypoints in c:\\anaconda\\lib\\site-packages (from jupyter-client>=6.1.12->ipykernel->jupyter) (0.4)\n",
      "Requirement already satisfied: platformdirs>=2.5 in c:\\anaconda\\lib\\site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->jupyter) (3.10.0)\n",
      "Requirement already satisfied: pywin32>=300 in c:\\anaconda\\lib\\site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->jupyter) (305.1)\n",
      "Requirement already satisfied: anyio<4,>=3.1.0 in c:\\anaconda\\lib\\site-packages (from jupyter-server<3,>=1.16.0->jupyterlab->jupyter) (3.5.0)\n",
      "Requirement already satisfied: websocket-client in c:\\anaconda\\lib\\site-packages (from jupyter-server<3,>=1.16.0->jupyterlab->jupyter) (0.58.0)\n",
      "Requirement already satisfied: pywinpty in c:\\anaconda\\lib\\site-packages (from jupyter-server<3,>=1.16.0->jupyterlab->jupyter) (2.0.10)\n",
      "Requirement already satisfied: jupyter-server-fileid<1,>=0.6.0 in c:\\anaconda\\lib\\site-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.9.0)\n",
      "Requirement already satisfied: ypy-websocket<0.9.0,>=0.8.2 in c:\\anaconda\\lib\\site-packages (from jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.8.2)\n",
      "Requirement already satisfied: y-py<0.6.0,>=0.5.3 in c:\\anaconda\\lib\\site-packages (from jupyter-ydoc~=0.2.3->jupyterlab->jupyter) (0.5.9)\n",
      "Requirement already satisfied: babel>=2.10 in c:\\anaconda\\lib\\site-packages (from jupyterlab-server~=2.19->jupyterlab->jupyter) (2.11.0)\n",
      "Requirement already satisfied: json5>=0.9.0 in c:\\anaconda\\lib\\site-packages (from jupyterlab-server~=2.19->jupyterlab->jupyter) (0.9.6)\n",
      "Requirement already satisfied: jsonschema>=4.18.0 in c:\\anaconda\\lib\\site-packages (from jupyterlab-server~=2.19->jupyterlab->jupyter) (4.19.2)\n",
      "Requirement already satisfied: requests>=2.31 in c:\\anaconda\\lib\\site-packages (from jupyterlab-server~=2.19->jupyterlab->jupyter) (2.31.0)\n",
      "Requirement already satisfied: notebook-shim>=0.2.3 in c:\\anaconda\\lib\\site-packages (from nbclassic->jupyterlab->jupyter) (0.2.3)\n",
      "Requirement already satisfied: fastjsonschema in c:\\anaconda\\lib\\site-packages (from nbformat->notebook->jupyter) (2.16.2)\n",
      "Requirement already satisfied: wcwidth in c:\\anaconda\\lib\\site-packages (from prompt-toolkit>=3.0.30->jupyter-console->jupyter) (0.2.5)\n",
      "Requirement already satisfied: argon2-cffi-bindings in c:\\anaconda\\lib\\site-packages (from argon2-cffi->notebook->jupyter) (21.2.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\anaconda\\lib\\site-packages (from beautifulsoup4->nbconvert->jupyter) (2.5)\n",
      "Requirement already satisfied: idna>=2.8 in c:\\anaconda\\lib\\site-packages (from anyio<4,>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab->jupyter) (3.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\anaconda\\lib\\site-packages (from anyio<4,>=3.1.0->jupyter-server<3,>=1.16.0->jupyterlab->jupyter) (1.3.1)\n",
      "Requirement already satisfied: pytz>=2015.7 in c:\\anaconda\\lib\\site-packages (from babel>=2.10->jupyterlab-server~=2.19->jupyterlab->jupyter) (2023.3.post1)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in c:\\anaconda\\lib\\site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel->jupyter) (0.8.3)\n",
      "Requirement already satisfied: attrs>=22.2.0 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (23.2.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (0.30.2)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (0.10.6)\n",
      "Requirement already satisfied: jupyter-events>=0.5.0 in c:\\anaconda\\lib\\site-packages (from jupyter-server-fileid<1,>=0.6.0->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.8.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\anaconda\\lib\\site-packages (from requests>=2.31->jupyterlab-server~=2.19->jupyterlab->jupyter) (2.0.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\anaconda\\lib\\site-packages (from requests>=2.31->jupyterlab-server~=2.19->jupyterlab->jupyter) (1.26.18)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\anaconda\\lib\\site-packages (from requests>=2.31->jupyterlab-server~=2.19->jupyterlab->jupyter) (2023.11.17)\n",
      "Requirement already satisfied: aiofiles<23,>=22.1.0 in c:\\anaconda\\lib\\site-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (22.1.0)\n",
      "Requirement already satisfied: aiosqlite<1,>=0.17.0 in c:\\anaconda\\lib\\site-packages (from ypy-websocket<0.9.0,>=0.8.2->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.18.0)\n",
      "Requirement already satisfied: cffi>=1.0.1 in c:\\anaconda\\lib\\site-packages (from argon2-cffi-bindings->argon2-cffi->notebook->jupyter) (1.16.0)\n",
      "Requirement already satisfied: executing in c:\\anaconda\\lib\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (0.8.3)\n",
      "Requirement already satisfied: asttokens in c:\\anaconda\\lib\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (2.0.5)\n",
      "Requirement already satisfied: pure-eval in c:\\anaconda\\lib\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter) (0.2.2)\n",
      "Requirement already satisfied: pycparser in c:\\anaconda\\lib\\site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook->jupyter) (2.21)\n",
      "Requirement already satisfied: python-json-logger>=2.0.4 in c:\\anaconda\\lib\\site-packages (from jupyter-events>=0.5.0->jupyter-server-fileid<1,>=0.6.0->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (2.0.7)\n",
      "Requirement already satisfied: pyyaml>=5.3 in c:\\anaconda\\lib\\site-packages (from jupyter-events>=0.5.0->jupyter-server-fileid<1,>=0.6.0->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (6.0.1)\n",
      "Requirement already satisfied: rfc3339-validator in c:\\anaconda\\lib\\site-packages (from jupyter-events>=0.5.0->jupyter-server-fileid<1,>=0.6.0->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.1.4)\n",
      "Requirement already satisfied: rfc3986-validator>=0.1.1 in c:\\anaconda\\lib\\site-packages (from jupyter-events>=0.5.0->jupyter-server-fileid<1,>=0.6.0->jupyter-server-ydoc~=0.8.0->jupyterlab->jupyter) (0.1.1)\n",
      "Requirement already satisfied: fqdn in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (1.5.1)\n",
      "Requirement already satisfied: isoduration in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (20.11.0)\n",
      "Requirement already satisfied: jsonpointer>1.13 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (2.1)\n",
      "Requirement already satisfied: uri-template in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (1.3.0)\n",
      "Requirement already satisfied: webcolors>=1.11 in c:\\anaconda\\lib\\site-packages (from jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (24.11.1)\n",
      "Requirement already satisfied: arrow>=0.15.0 in c:\\anaconda\\lib\\site-packages (from isoduration->jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (1.3.0)\n",
      "Requirement already satisfied: types-python-dateutil>=2.8.10 in c:\\anaconda\\lib\\site-packages (from arrow>=0.15.0->isoduration->jsonschema>=4.18.0->jupyterlab-server~=2.19->jupyterlab->jupyter) (2.9.0.20241206)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install face_recognition opencv-python dlib numpy pillow matplotlib jupyter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gerekli kütüphaneleri import et\n",
    "import face_recognition\n",
    "import cv2\n",
    "import numpy as np\n",
    "from datetime import datetime\n",
    "import os\n",
    "import logging\n",
    "import dlib\n",
    "from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "import bz2\n",
    "import shutil\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "# Logging ayarları\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model dosyası mevcut.\n"
     ]
    }
   ],
   "source": [
    "def download_model(model_path: str) -> bool:\n",
    "    \"\"\"Model dosyasını indirir ve belirtilen konuma kaydeder\"\"\"\n",
    "    try:\n",
    "        url = 'http://dlib.net/files/shape_predictor_68_face_landmarks.dat.bz2'\n",
    "        temp_path = f\"{model_path}.bz2\"\n",
    "        \n",
    "        print(f\"Model indiriliyor: {url}\")\n",
    "        response = requests.get(url, stream=True)\n",
    "        response.raise_for_status()\n",
    "        \n",
    "        with open(temp_path, 'wb') as f:\n",
    "            for chunk in response.iter_content(chunk_size=8192):\n",
    "                if chunk:\n",
    "                    f.write(chunk)\n",
    "        \n",
    "        print(\"Model dosyası çıkartılıyor...\")\n",
    "        with bz2.open(temp_path, 'rb') as source, open(model_path, 'wb') as dest:\n",
    "            shutil.copyfileobj(source, dest)\n",
    "        \n",
    "        os.remove(temp_path)\n",
    "        print(f\"Model başarıyla indirildi: {model_path}\")\n",
    "        return True\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Model indirme hatası: {str(e)}\")\n",
    "        return False\n",
    "\n",
    "def check_and_download_model():\n",
    "    try:\n",
    "        models_dir = os.path.join(os.getcwd(), 'models')\n",
    "        os.makedirs(models_dir, exist_ok=True)\n",
    "        \n",
    "        model_path = os.path.join(models_dir, 'shape_predictor_68_face_landmarks.dat')\n",
    "        if not os.path.exists(model_path):\n",
    "            print(\"Model dosyası bulunamadı, indiriliyor...\")\n",
    "            if not download_model(model_path):\n",
    "                raise RuntimeError(\"Model indirilemedi!\")\n",
    "        else:\n",
    "            print(\"Model dosyası mevcut.\")\n",
    "            \n",
    "        return model_path\n",
    "    except Exception as e:\n",
    "        print(f\"Hata: {str(e)}\")\n",
    "        return None\n",
    "\n",
    "# Model kontrolü yap\n",
    "model_path = check_and_download_model()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class FaceRecognition:\n",
    "    def __init__(self):\n",
    "        self.known_face_encodings = []\n",
    "        self.known_face_names = []\n",
    "        \n",
    "    def load_known_face(self, image_path, person_name):\n",
    "        \"\"\"Bilinen bir yüzü yükle\"\"\"\n",
    "        try:\n",
    "            image = face_recognition.load_image_file(image_path)\n",
    "            face_encodings = face_recognition.face_encodings(image)\n",
    "            \n",
    "            if face_encodings:\n",
    "                self.known_face_encodings.append(face_encodings[0])\n",
    "                self.known_face_names.append(person_name)\n",
    "                print(f\"{person_name} başarıyla eklendi.\")\n",
    "                return True\n",
    "            else:\n",
    "                print(f\"Uyarı: {image_path} dosyasında yüz bulunamadı!\")\n",
    "                return False\n",
    "                \n",
    "        except Exception as e:\n",
    "            print(f\"Hata: {str(e)}\")\n",
    "            return False\n",
    "    \n",
    "    def recognize_faces(self, image_path):\n",
    "        \"\"\"Görüntüdeki yüzleri tanı\"\"\"\n",
    "        try:\n",
    "            image = face_recognition.load_image_file(image_path)\n",
    "            face_locations = face_recognition.face_locations(image)\n",
    "            face_encodings = face_recognition.face_encodings(image, face_locations)\n",
    "            \n",
    "            face_names = []\n",
    "            for face_encoding in face_encodings:\n",
    "                matches = face_recognition.compare_faces(\n",
    "                    self.known_face_encodings,\n",
    "                    face_encoding,\n",
    "                    tolerance=0.6\n",
    "                )\n",
    "                name = \"Bilinmeyen\"\n",
    "                \n",
    "                if True in matches:\n",
    "                    first_match_index = matches.index(True)\n",
    "                    name = self.known_face_names[first_match_index]\n",
    "                    \n",
    "                face_names.append(name)\n",
    "            \n",
    "            # Görüntüyü göster\n",
    "            self.display_results(image, face_locations, face_names)\n",
    "            return face_names\n",
    "            \n",
    "        except Exception as e:\n",
    "            print(f\"Hata: {str(e)}\")\n",
    "            return []\n",
    "    \n",
    "    def display_results(self, image, face_locations, face_names):\n",
    "        \"\"\"Sonuçları görselleştir\"\"\"\n",
    "        image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)\n",
    "        \n",
    "        for (top, right, bottom, left), name in zip(face_locations, face_names):\n",
    "            # Yüz çerçevesi\n",
    "            cv2.rectangle(image_rgb, (left, top), (right, bottom), (0, 255, 0), 2)\n",
    "            \n",
    "            # İsim etiketi\n",
    "            cv2.putText(\n",
    "                image_rgb,\n",
    "                name,\n",
    "                (left, top - 10),\n",
    "                cv2.FONT_HERSHEY_SIMPLEX,\n",
    "                0.75,\n",
    "                (0, 255, 0),\n",
    "                2\n",
    "            )\n",
    "        \n",
    "        plt.figure(figsize=(12, 8))\n",
    "        plt.imshow(image_rgb)\n",
    "        plt.axis('off')\n",
    "        plt.show()\n",
    "\n",
    "# Yüz tanıma nesnesini oluştur\n",
    "face_recognizer = FaceRecognition()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yusuf başarıyla eklendi.\n",
      "ömer başarıyla eklendi.\n"
     ]
    }
   ],
   "source": [
    "known_faces_dir = 'known_faces'\n",
    "os.makedirs(known_faces_dir, exist_ok=True)\n",
    "\n",
    "if os.path.exists(known_faces_dir):\n",
    "    for filename in os.listdir(known_faces_dir):\n",
    "        if filename.endswith(('.jpg', '.jpeg', '.png')):\n",
    "            name = os.path.splitext(filename)[0]\n",
    "            image_path = os.path.join(known_faces_dir, filename)\n",
    "            face_recognizer.load_known_face(image_path, name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_person(image_path, name):\n",
    "    if face_recognizer.load_known_face(image_path, name):\n",
    "        # Fotoğrafı known_faces klasörüne kopyala\n",
    "        extension = os.path.splitext(image_path)[1]\n",
    "        new_path = os.path.join(known_faces_dir, f\"{name}{extension}\")\n",
    "        shutil.copy2(image_path, new_path)\n",
    "        print(f\"{name} başarıyla eklendi ve fotoğraf kaydedildi.\")\n",
    "    else:\n",
    "        print(\"Kişi eklenemedi.\")\n",
    "\n",
    "# Örnek kullanım:\n",
    "# add_person('path/to/photo.jpg', 'Ahmet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_recognition(image_path):\n",
    "    if os.path.exists(image_path):\n",
    "        detected_faces = face_recognizer.recognize_faces(image_path)\n",
    "        print(f\"\\nTanınan yüzler: {detected_faces}\")\n",
    "    else:\n",
    "        print(f\"Hata: {image_path} bulunamadı!\")\n",
    "\n",
    "# Örnek kullanım:\n",
    "# test_recognition('test.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def test_webcam():\n",
    "    cap = cv2.VideoCapture(0)\n",
    "    \n",
    "    try:\n",
    "        while True:\n",
    "            ret, frame = cap.read()\n",
    "            if not ret:\n",
    "                break\n",
    "                \n",
    "            # BGR'den RGB'ye dönüştür\n",
    "            rgb_frame = frame[:, :, ::-1]\n",
    "            \n",
    "            # Yüzleri bul ve tanı\n",
    "            face_locations = face_recognition.face_locations(rgb_frame)\n",
    "            face_encodings = face_recognition.face_encodings(rgb_frame, face_locations)\n",
    "            \n",
    "            for (top, right, bottom, left), face_encoding in zip(face_locations, face_encodings):\n",
    "                matches = face_recognition.compare_faces(\n",
    "                    face_recognizer.known_face_encodings,\n",
    "                    face_encoding\n",
    "                )\n",
    "                name = \"Bilinmeyen\"\n",
    "                \n",
    "                if True in matches:\n",
    "                    first_match_index = matches.index(True)\n",
    "                    name = face_recognizer.known_face_names[first_match_index]\n",
    "                \n",
    "                # Yüzü çerçevele\n",
    "                cv2.rectangle(frame, (left, top), (right, bottom), (0, 255, 0), 2)\n",
    "                cv2.putText(\n",
    "                    frame,\n",
    "                    name,\n",
    "                    (left, top - 10),\n",
    "                    cv2.FONT_HERSHEY_SIMPLEX,\n",
    "                    0.75,\n",
    "                    (0, 255, 0),\n",
    "                    2\n",
    "                )\n",
    "            \n",
    "            # Görüntüyü göster\n",
    "            cv2.imshow('Webcam', frame)\n",
    "            \n",
    "            # 'q' tuşuna basılırsa çık\n",
    "            if cv2.waitKey(1) & 0xFF == ord('q'):\n",
    "                break\n",
    "                \n",
    "    finally:\n",
    "        cap.release()\n",
    "        cv2.destroyAllWindows()\n",
    "\n",
    "# Webcam testi için bu satırı çalıştırın:\n",
    "# test_webcam()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:__main__:Hata: 'FaceRecognitionApp' object has no attribute 'status_var'\n"
     ]
    }
   ],
   "source": [
    "import face_recognition\n",
    "import cv2\n",
    "import numpy as np\n",
    "from datetime import datetime\n",
    "import os\n",
    "import logging\n",
    "import dlib\n",
    "from PIL import Image\n",
    "import matplotlib.pyplot as plt\n",
    "import requests\n",
    "import bz2\n",
    "import shutil\n",
    "import tkinter as tk\n",
    "from tkinter import ttk, filedialog, messagebox\n",
    "from PIL import Image, ImageTk\n",
    "\n",
    "class FaceRecognitionApp:\n",
    "    def __init__(self, root):\n",
    "        self.root = root\n",
    "        self.root.title(\"Yüz Tanıma Sistemi\")\n",
    "        self.root.geometry(\"800x600\")\n",
    "        \n",
    "        # Logging ayarları\n",
    "        logging.basicConfig(level=logging.INFO)\n",
    "        self.logger = logging.getLogger(__name__)\n",
    "        \n",
    "        # Model kontrolü\n",
    "        self.check_and_download_model()\n",
    "        \n",
    "        # Yüz tanıma sınıfını başlat\n",
    "        self.face_recognizer = FaceRecognition()\n",
    "        \n",
    "        # GUI bileşenlerini oluştur\n",
    "        self.create_gui()\n",
    "        \n",
    "        # Known faces klasörünü oluştur\n",
    "        self.known_faces_dir = 'known_faces'\n",
    "        os.makedirs(self.known_faces_dir, exist_ok=True)\n",
    "        \n",
    "        # Mevcut yüzleri yükle\n",
    "        self.load_known_faces()\n",
    "\n",
    "    def create_gui(self):\n",
    "        # Ana frame\n",
    "        main_frame = ttk.Frame(self.root, padding=\"10\")\n",
    "        main_frame.grid(row=0, column=0, sticky=(tk.W, tk.E, tk.N, tk.S))\n",
    "        \n",
    "        # Kamera görüntüsü için canvas\n",
    "        self.camera_canvas = tk.Canvas(main_frame, width=640, height=480)\n",
    "        self.camera_canvas.grid(row=0, column=0, columnspan=2, pady=10)\n",
    "        \n",
    "        # Butonlar için frame\n",
    "        button_frame = ttk.Frame(main_frame)\n",
    "        button_frame.grid(row=1, column=0, columnspan=2, pady=10)\n",
    "        \n",
    "        # Butonlar\n",
    "        ttk.Button(button_frame, text=\"Kişi Ekle\", command=self.add_person_dialog).pack(side=tk.LEFT, padx=5)\n",
    "        self.camera_button = ttk.Button(button_frame, text=\"Kamerayı Başlat\", command=self.toggle_camera)\n",
    "        self.camera_button.pack(side=tk.LEFT, padx=5)\n",
    "        ttk.Button(button_frame, text=\"Fotoğraftan Tanı\", command=self.recognize_from_photo).pack(side=tk.LEFT, padx=5)\n",
    "        \n",
    "        # Durum çubuğu\n",
    "        self.status_var = tk.StringVar()\n",
    "        self.status_var.set(\"Hazır\")\n",
    "        status_label = ttk.Label(main_frame, textvariable=self.status_var)\n",
    "        status_label.grid(row=2, column=0, columnspan=2, pady=5)\n",
    "        \n",
    "        # Kamera değişkenleri\n",
    "        self.camera_active = False\n",
    "        self.cap = None\n",
    "\n",
    "    def add_person_dialog(self):\n",
    "        dialog = tk.Toplevel(self.root)\n",
    "        dialog.title(\"Kişi Ekle\")\n",
    "        dialog.geometry(\"300x150\")\n",
    "        \n",
    "        ttk.Label(dialog, text=\"İsim:\").pack(pady=5)\n",
    "        name_entry = ttk.Entry(dialog)\n",
    "        name_entry.pack(pady=5)\n",
    "        \n",
    "        def select_photo():\n",
    "            file_path = filedialog.askopenfilename(\n",
    "                filetypes=[(\"Image files\", \"*.jpg *.jpeg *.png\")]\n",
    "            )\n",
    "            if file_path:\n",
    "                try:\n",
    "                    self.add_person(file_path, name_entry.get())\n",
    "                    dialog.destroy()\n",
    "                except Exception as e:\n",
    "                    messagebox.showerror(\"Hata\", str(e))\n",
    "        \n",
    "        ttk.Button(dialog, text=\"Fotoğraf Seç\", command=select_photo).pack(pady=10)\n",
    "\n",
    "    def add_person(self, image_path, name):\n",
    "        if not name:\n",
    "            raise ValueError(\"Lütfen bir isim girin!\")\n",
    "            \n",
    "        if self.face_recognizer.load_known_face(image_path, name):\n",
    "            # Fotoğrafı known_faces klasörüne kopyala\n",
    "            extension = os.path.splitext(image_path)[1]\n",
    "            new_path = os.path.join(self.known_faces_dir, f\"{name}{extension}\")\n",
    "            shutil.copy2(image_path, new_path)\n",
    "            self.status_var.set(f\"{name} başarıyla eklendi\")\n",
    "            messagebox.showinfo(\"Başarılı\", f\"{name} başarıyla eklendi\")\n",
    "        else:\n",
    "            raise ValueError(\"Fotoğrafta yüz bulunamadı!\")\n",
    "\n",
    "    def toggle_camera(self):\n",
    "        if not self.camera_active:\n",
    "            self.cap = cv2.VideoCapture(0)\n",
    "            if not self.cap.isOpened():\n",
    "                messagebox.showerror(\"Hata\", \"Kamera açılamadı!\")\n",
    "                return\n",
    "                \n",
    "            self.camera_active = True\n",
    "            self.camera_button.configure(text=\"Kamerayı Durdur\")\n",
    "            self.update_camera()\n",
    "        else:\n",
    "            self.camera_active = False\n",
    "            self.camera_button.configure(text=\"Kamerayı Başlat\")\n",
    "            if self.cap:\n",
    "                self.cap.release()\n",
    "\n",
    "    def update_camera(self):\n",
    "        if self.camera_active:\n",
    "            ret, frame = self.cap.read()\n",
    "            if ret:\n",
    "                # Yüz tanıma işlemi\n",
    "                rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)\n",
    "                face_locations = face_recognition.face_locations(rgb_frame)\n",
    "                face_encodings = face_recognition.face_encodings(rgb_frame, face_locations)\n",
    "                \n",
    "                for (top, right, bottom, left), face_encoding in zip(face_locations, face_encodings):\n",
    "                    matches = face_recognition.compare_faces(\n",
    "                        self.face_recognizer.known_face_encodings,\n",
    "                        face_encoding\n",
    "                    )\n",
    "                    name = \"Bilinmeyen\"\n",
    "                    \n",
    "                    if True in matches:\n",
    "                        first_match_index = matches.index(True)\n",
    "                        name = self.face_recognizer.known_face_names[first_match_index]\n",
    "                    \n",
    "                    # Yüzü çerçevele\n",
    "                    cv2.rectangle(frame, (left, top), (right, bottom), (0, 255, 0), 2)\n",
    "                    cv2.putText(\n",
    "                        frame,\n",
    "                        name,\n",
    "                        (left, top - 10),\n",
    "                        cv2.FONT_HERSHEY_SIMPLEX,\n",
    "                        0.75,\n",
    "                        (0, 255, 0),\n",
    "                        2\n",
    "                    )\n",
    "                \n",
    "                # Görüntüyü canvas'a aktar\n",
    "                self.photo = ImageTk.PhotoImage(image=Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)))\n",
    "                self.camera_canvas.create_image(0, 0, image=self.photo, anchor=tk.NW)\n",
    "            \n",
    "            self.root.after(10, self.update_camera)\n",
    "\n",
    "    def recognize_from_photo(self):\n",
    "        file_path = filedialog.askopenfilename(\n",
    "            filetypes=[(\"Image files\", \"*.jpg *.jpeg *.png\")]\n",
    "        )\n",
    "        if file_path:\n",
    "            detected_faces = self.face_recognizer.recognize_faces(file_path)\n",
    "            if detected_faces:\n",
    "                messagebox.showinfo(\"Sonuç\", f\"Tanınan yüzler: {', '.join(detected_faces)}\")\n",
    "            else:\n",
    "                messagebox.showinfo(\"Sonuç\", \"Yüz bulunamadı veya tanınamadı\")\n",
    "\n",
    "    def load_known_faces(self):\n",
    "        if os.path.exists(self.known_faces_dir):\n",
    "            for filename in os.listdir(self.known_faces_dir):\n",
    "                if filename.endswith(('.jpg', '.jpeg', '.png')):\n",
    "                    name = os.path.splitext(filename)[0]\n",
    "                    image_path = os.path.join(self.known_faces_dir, filename)\n",
    "                    self.face_recognizer.load_known_face(image_path, name)\n",
    "\n",
    "    def check_and_download_model(self):\n",
    "        try:\n",
    "            models_dir = os.path.join(os.getcwd(), 'models')\n",
    "            os.makedirs(models_dir, exist_ok=True)\n",
    "            \n",
    "            model_path = os.path.join(models_dir, 'shape_predictor_68_face_landmarks.dat')\n",
    "            if not os.path.exists(model_path):\n",
    "                self.status_var.set(\"Model indiriliyor...\")\n",
    "                if not self.download_model(model_path):\n",
    "                    raise RuntimeError(\"Model indirilemedi!\")\n",
    "            \n",
    "            self.status_var.set(\"Model hazır\")\n",
    "            return model_path\n",
    "        except Exception as e:\n",
    "            self.logger.error(f\"Hata: {str(e)}\")\n",
    "            messagebox.showerror(\"Hata\", str(e))\n",
    "            return None\n",
    "\n",
    "    def download_model(self, model_path):\n",
    "        try:\n",
    "            url = 'http://dlib.net/files/shape_predictor_68_face_landmarks.dat.bz2'\n",
    "            temp_path = f\"{model_path}.bz2\"\n",
    "            \n",
    "            response = requests.get(url, stream=True)\n",
    "            response.raise_for_status()\n",
    "            \n",
    "            with open(temp_path, 'wb') as f:\n",
    "                for chunk in response.iter_content(chunk_size=8192):\n",
    "                    if chunk:\n",
    "                        f.write(chunk)\n",
    "            \n",
    "            with bz2.open(temp_path, 'rb') as source, open(model_path, 'wb') as dest:\n",
    "                shutil.copyfileobj(source, dest)\n",
    "            \n",
    "            os.remove(temp_path)\n",
    "            return True\n",
    "            \n",
    "        except Exception as e:\n",
    "            self.logger.error(f\"Model indirme hatası: {str(e)}\")\n",
    "            return False\n",
    "\n",
    "class FaceRecognition:\n",
    "    def __init__(self):\n",
    "        self.known_face_encodings = []\n",
    "        self.known_face_names = []\n",
    "        \n",
    "    def load_known_face(self, image_path, person_name):\n",
    "        try:\n",
    "            image = face_recognition.load_image_file(image_path)\n",
    "            face_encodings = face_recognition.face_encodings(image)\n",
    "            \n",
    "            if face_encodings:\n",
    "                self.known_face_encodings.append(face_encodings[0])\n",
    "                self.known_face_names.append(person_name)\n",
    "                return True\n",
    "            return False\n",
    "                \n",
    "        except Exception as e:\n",
    "            logging.error(f\"Hata: {str(e)}\")\n",
    "            return False\n",
    "    \n",
    "    def recognize_faces(self, image_path):\n",
    "        try:\n",
    "            image = face_recognition.load_image_file(image_path)\n",
    "            face_locations = face_recognition.face_locations(image)\n",
    "            face_encodings = face_recognition.face_encodings(image, face_locations)\n",
    "            \n",
    "            face_names = []\n",
    "            for face_encoding in face_encodings:\n",
    "                matches = face_recognition.compare_faces(\n",
    "                    self.known_face_encodings,\n",
    "                    face_encoding,\n",
    "                    tolerance=0.6\n",
    "                )\n",
    "                name = \"Bilinmeyen\"\n",
    "                \n",
    "                if True in matches:\n",
    "                    first_match_index = matches.index(True)\n",
    "                    name = self.known_face_names[first_match_index]\n",
    "                    \n",
    "                face_names.append(name)\n",
    "            \n",
    "            return face_names\n",
    "            \n",
    "        except Exception as e:\n",
    "            logging.error(f\"Hata: {str(e)}\")\n",
    "            return []\n",
    "\n",
    "def main():\n",
    "    root = tk.Tk()\n",
    "    app = FaceRecognitionApp(root)\n",
    "    root.mainloop()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
